File: degree-of-an-array/degree-of-an-array.java
// n = nums.length
// Time: O(n)
// Space: O(n)
class Solution {
    public int findShortestSubArray(int[] nums) {
        Map<Integer, Integer> map = new HashMap<>();
        for (int num : nums)
            map.put(num, map.getOrDefault(num, 0) + 1);
        
        int degree = 0;
        for (int key : map.keySet())
            degree = Math.max(degree, map.get(key));
        
        map.clear();
        int minLen = Integer.MAX_VALUE;
        for (int i = 0, j = 0; i < nums.length; i++) {
            map.put(nums[i], map.getOrDefault(nums[i], 0) + 1);                
            while (degCount(map, degree) == 1) {
                minLen = Math.min(minLen, i - j + 1);
                map.put(nums[j], map.get(nums[j]) - 1);
                if (map.get(nums[j]) == 0)
                    map.remove(nums[j]);
                j++;
            }
        }
        
        return minLen;
    }
    
    private int degCount(Map<Integer, Integer> map, int degree) {
        int count = 0;
        for (int key : map.keySet()) {
            if (map.get(key) == degree)
                count++;
        }
        return count;
    }
}

File: degree-of-an-array/degree-of-an-array.cpp
class Solution {
public:
  int findShortestSubArray(vector<int>& nums) {
    unordered_map<int, pair<int, int>> fl;
    unordered_map<int, int> cnt;
    int deg = 0;
     
    for(int i=0;i<nums.size();i++){
      deg = max(deg, ++cnt[nums[i]]);
      fl[nums[i]].second = i;
    }
   
    for(int i=nums.size()-1;i>=0;i--)
      fl[nums[i]].first = i;
     
    int ans = nums.size();
    for(auto k: cnt){
      if(k.second != deg)
        continue;
       
      int s = fl[k.first].first;
      int e = fl[k.first].second;
      ans = min(ans, e-s+1);
    }
     
    return ans;
  }
};